code-dot-org/code-dot-org
# LTI account linking controller and service

The first PR to start implementing account linking for LTI.

- Adds a controller to handle the new account linking pages and routes. This includes a route for linking a new LTI login to an existing email/password account.
- Adds a service for account linking. The main LTI service is getting large, and there will be more code once we start implementing linking LTI logins to Oauth accounts, so I figured this deserves its own service.
- Adds unstyled placeholders for the new LTI registration landing page and account linking page.
- All the new routes/pages are gated behind a DOTD flag to keep them disabled until we're ready to ship.

## Links

[JIRA](https://codedotorg.atlassian.net/browse/P20-908)
[LTI tech spec](https://docs.google.com/document/d/1tTwAF40cnn2oUvK2CmRO-a5qYXkCR5rR5gBJS2-zV_U/edit#heading=h.y0j9qhlwm35z)

## Testing story

To test this locally, set enable the DCDO flag:
```
DCDO.set('lti_account_linking_enabled', true)
```

Then make sure you have an existing user with an email login to test with. Launch from a new user in the LMS (or just delete any existing LTI auth_options for that user). Click "I already have a Code.org account" and then enter the creds for your existing email account. Your LTI auth option should now be attached to the email account, and you will be logged in and redirected to the homepage.

Unit Tests:
```
bundle exec spring testunit test/controllers/lti/v1/users_controller_test.rb
  4/4: [====================================================] 100% Time: 00:00:01, Time: 00:00:01

Finished in 1.12101s
4 tests, 9 assertions, 0 failures, 0 errors, 0 skips
```

```
bundle exec spring testunit test/lib/services/lti/account_linker_test.rb
  1/1: [====================================================] 100% Time: 00:00:00, Time: 00:00:00

Finished in 0.25297s
1 tests, 4 assertions, 0 failures, 0 errors, 0 skips
```



## Deployment strategy

## Follow-up work

This PR just lays the groundwork for the rest of the LTI account linking feature. Not intended to be prod ready yet, so it's behind a DCDO flag.

## Privacy



## Security



## Caching

## PR Checklist:
